test/spec/modules/33acrossAnalyticsAdapter_spec.js

```diff
@@ -0,0 +1,2368 @@
+              provider: '33across',
+              options: {
+                pid: 'test-pid',
+                endpoint: 'http://test-endpoint',
+                timeout: value
+              },
+            });
+            analyticsAdapter.disableAnalytics();
+
+            assert.called(log.info);
+            assert.calledWithExactly(log.info.getCall(index), 'Invalid timeout provided for "options.timeout". Using default timeout of 3000ms.');
+          });
+        });
+      });
+
+      it('logs any received GAM slotRenderEnded event', function() {
+        analyticsAdapter.enableAnalytics({
+          provider: '33across',
+          options: {
+            pid: 'test-pid',
+            endpoint: faker.internet.url(),
+          },
+        });
+
+        performAuctionWithMissingBidWon();
+
+        window.googletag.cmd.forEach(fn => fn());
+
+        const { gam } = getMockEvents();
+        const [ gEvent ] = gam.slotRenderEnded;
+
+        mockGpt.emitEvent('slotRenderEnded', gEvent);
+
+        assert.calledWithExactly(log.info, 'slotRenderEnded', `adUnitPath: ${gEvent.slot.getAdUnitPath()}`, `slotElementId: ${gEvent.slot.getSlotElementId()}`);
+      });
+    });
+  });
+
+  describe('Event Handling', function() {
+    beforeEach(function() {
+      this.defaultTimeout = 3000;
+      this.enableAnalytics = (options) => {
+        analyticsAdapter.enableAnalytics({
+          provider: '33across',
+          options: {
+            endpoint: 'http://test-endpoint',
+            pid: 'test-pid',
+            timeout: this.defaultTimeout,
+            ...options
+          },
+        });
+      }
+    });
+
+    context('when an auction is complete', function() {
+      context('and the AnalyticsReport is sent successfully to the given endpoint', function() {
+        it('it logs an info message', function() {
+          const endpoint = faker.internet.url();
+          this.enableAnalytics({ endpoint });
+
+          sandbox.stub(navigator, 'sendBeacon').returns(true);
+
+          performStandardAuction();
+          sandbox.clock.tick(this.defaultTimeout + 1);
+
+          assert.calledOnce(navigator.sendBeacon);
+          assert.calledWith(navigator.sendBeacon, endpoint, sinon.match.string);
+          assert.calledWith(log.info, `Analytics report sent to ${endpoint}`, sinon.match.object);
+
+          navigator.sendBeacon.restore();
+        });
+      });
+    });
+
+    context('when an error occurs while sending the AnalyticsReport', function() {
+      it('logs an error', function() {
+        this.enableAnalytics();
+
+        sandbox.stub(navigator, 'sendBeacon').returns(false);
+
+        performStandardAuction();
+        sandbox.clock.tick(this.defaultTimeout + 1);
+
+        assert.calledOnce(navigator.sendBeacon);
+        assert.calledOnce(log.error);
+        assert.calledWith(log.error, 'Analytics report exceeded User-Agent data limits and was not sent.', sinon.match.object);
```

Comment:
I think there's a function like `calledOnceWith`, that way you could remove combine the last two assertions.

Reply:
I like having the multiple assertions so that you can more quickly and precisely locate the failure point if/when the test fails.

Reply:
got it.. 

I am just mentioning this limit on the number of assertions because at some point, during guideline discussions we decided that each spec should have max 2 assertions, I think we do do have a doc that talks about it.. I'll try to find it.

Technically, the assertion should be related to the title of the spec. So `assert.calledOnce(navigator.sendBeacon);` belongs more to a spec which title is related, for instance something like `it('makes a single request') ... `  

just wanted to be less strict and at least keep the assertions to the minimum, and mention that you always have the option to split the spec and keep the title-assertion relation.

Reply:
here is the guidelines doc that talks about assertions.. it was created for SIC but we try to also apply the same guidelines on other client-side projects.

https://confluence.internal.33across.com/display/ENG/Creating+Unit+Tests+for+Pre-Existing+Code#CreatingUnitTestsforPreExistingCode-_8pjkc79qg4h